hal/armv7a/zynq7000/zynq.c

```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
+				((!!clkact1) << 1) | ((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8);
+			break;
+
+		case pctl_ctrl_can_clk:
+			*(zynq_common.slcr + slcr_can_clk_ctrl) = (*(zynq_common.slcr + slcr_can_clk_ctrl) & ~0x03f03f33) | (!!clkact0) |
+				((!!clkact1) << 1) | ((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8) | ((divisor1 & 0x3f) << 20);
+			break;
+
+		default:
+			_zynq_slcrLock();
+			return -1;
+	}
+
+	_zynq_slcrLock();
+
+	return 0;
+}
+
+
+static int _zynq_getDevClk(int dev, char *divisor0, char *divisor1, char *srcsel, char *clkact0, char *clkact1)
+{
+	u32 id;
+	u32 val = 0;
+
+	switch (dev) {
+		case pctl_ctrl_usb0_clk:
+		case pctl_ctrl_usb1_clk:
+			id = dev - pctl_ctrl_usb0_clk;
+			val = *(zynq_common.slcr + slcr_usb0_clk_ctrl + id);
+			*srcsel = (val >> 4) & 0x7;
+			*clkact0 = *clkact1 = *divisor0 = *divisor1 = 0;
+			break;
+
+		case pctl_ctrl_gem0_rclk:
+		case pctl_ctrl_gem1_rclk:
+			id = dev - pctl_ctrl_gem0_rclk;
+			val = *(zynq_common.slcr + slcr_gem0_rclk_ctrl + id);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x1;
+			*clkact1 = *divisor0 = *divisor1 = 0;
+			return 0;
+
+		case pctl_ctrl_gem0_clk:
+		case pctl_ctrl_gem1_clk:
+			id = dev - pctl_ctrl_gem0_clk;
+			val = *(zynq_common.slcr + slcr_gem0_clk_ctrl + id);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x7;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = (val >> 20) & 0x3f;
+			*clkact1 = 0;
+			break;
+
+		case pctl_ctrl_smc_clk:
+			val = *(zynq_common.slcr + slcr_smc_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*clkact1 = *divisor1 = 0;
+			break;
+
+		case pctl_ctrl_lqspi_clk:
+			val = *(zynq_common.slcr + slcr_lqspi_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*clkact1 = *divisor1 = 0;
+			break;
+
+		case pctl_ctrl_sdio_clk:
+			val = *(zynq_common.slcr + slcr_sdio_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = 0;
+			break;
+
+		case pctl_ctrl_uart_clk:
+			val = *(zynq_common.slcr + slcr_uart_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = 0;
+			break;
+
+		case pctl_ctrl_spi_clk:
+			val = *(zynq_common.slcr + slcr_spi_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = 0;
+			break;
+
+		case pctl_ctrl_can_clk:
+			val = *(zynq_common.slcr + slcr_can_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = (val >> 20) & 0x3f;
+			break;
+
+		default:
+			return -1;
+	}
+
+	return 0;
+}
+
+
+static int _zynq_setMioClk(int mio, char ref0, char mux0, char ref1, char mux1)
+{
+	if (mio == pctl_ctrl_can_mioclk) {
```

Comment:
Hmm, what is the point of even massing `int mio` argument? I suppose other clock are TODO? In that case comment about that would be nice

Reply:
`MioClk` can be set only to `can`. To more readable I can add `err = -1` in else statement. What do you think?

Reply:
I think there is no point passing `mio` argument to this function, if only one special value is accepted. Better to call this (along with `platformctl`) along lines `setMioCanClk` and drop `mio` argument. Same applies to `_zynq_getMioClk`

Reply:
Done